cirq/study/flattener.py

```diff
@@ -0,0 +1,275 @@
+                symbols, or expressions to other symbols or values.  Only sympy
+                expressions and symbols not specified in `param_dict` will be
+                flattened.
+            new_param_names: An iterator of parameter names to replace each
+                expression with.  It may be infinite but there must be no
+                repetitions.  By default the parameter names are x0, x1, x2, ...
+        """
+        if hasattr(self, '_param_name_gen'):
+            # Has already been initialized
+            return
+        if isinstance(param_dict, ParamResolver):
+            params = param_dict.param_dict
+        else:
+            params = param_dict if param_dict else {}
+        symbol_params = {
+            _ensure_not_str(param): _ensure_not_str(val)
+            for param, val in params.items()
+        }
+        super().__init__(symbol_params)
+        self._param_name_gen = iter(new_param_names)
+        self.new_param_names = new_param_names
+        self._taken_symbols = set(self.param_dict.values())
+
+    @staticmethod
+    def default_param_names(prefix: str = 'x', start: int = 0) -> Iterator[str]:
+        """Generates an infinite sequence of parameter names with the format
+        '<prefix><index>' with no repetitions.  E.g. x0, x1, x2, ...
+
+        Args:
+            prefix: The prefix string of the parameter name.
+            start: The index of the first parameter name generated.
+
+        Returns:
+            An infinite iterator of parameter names.
+        """
+        for i in itertools.count(start):
+            yield '{}{}'.format(prefix, i)
+
+    def _next_name(self) -> str:
+        try:
+            return next(self._param_name_gen)
+        except StopIteration:
+            self._param_name_gen = self.default_param_names()
+            return next(self._param_name_gen)
+
+    def value_of(self, value: Union[sympy.Basic, float, str]
+                ) -> Union[sympy.Basic, float]:
+        """Resolves a symbol or expression to a new symbol unique to that value.
+
+        If value is a float, returns it.
+        If this `ParamFlattener` was initialized with a `param_dict` and `value`
+        is a key, returns `param_dict[value]`.
+        Otherwise return a symbol unique to the given value.
+
+        Args:
+            value: The sympy.Symbol, sympy expression, name, or float to resolve
+                to a unique symbol or float.
+
+        Returns:
+            The unique symbol or value of the parameter as resolved by this
+            resolver.
+        """
+        if isinstance(value, (int, float)):
+            return value
+        if isinstance(value, str):
+            value = sympy.Symbol(value)
+        out = self.param_dict.get(value, None)
+        if out is not None:
+            return out
+        # Create a new symbol
+        symbol = sympy.Symbol(self._next_name())
+        # Ensure the symbol hasn't already been used
```

Comment:
I'd consider moving this into _next_name since then it makes the guarantee of _next_name that it is unique.

Reply:
Done.